src/gsi/general_read_gfsatm.f90

```diff
@@ -2161,353 +2267,432 @@ subroutine general_read_gfsatm_nc(grd,sp_a,filename,uvflag,vordivflag,zflag, &
+          if ( icount == icm ) then
+             call general_reload(grd,g_z,g_ps,g_tv,g_vor,g_div,g_u,g_v,g_q,g_oz,g_cwmr, &
+                  icount,iflag,ilev,work,uvflag,vordivflag)
+          endif
+       end do
+
+       do k=1,nlevs
+          icount=icount+1
+          iflag(icount)=10
+          ilev(icount)=k
+          kr = levs+1-k ! netcdf is top to bottom, need to flip
+
+          if (mype==mype_use(icount)) then
+             call read_vardata(filges, 'clwmr', rwork3d0, nslice=kr, slicedim=3)
+             call read_vardata(filges, 'icmr', rwork3d1, nslice=kr, slicedim=3)
+             ! Cloud condensate mixing ratio.
+             rwork2d = rwork3d0(:,:,1)+rwork3d1(:,:,1)
+             if ( diff_res ) then
+                grid_b=rwork2d
+                vector(1)=.false.
+                call fill2_ns(grid_b,grid_c(:,:,1),latb+2,lonb)
+                call g_egrid2agrid(p_high,grid_c,grid2,1,1,vector)
+                do kk=1,grd%itotsub
+                   i=grd%ltosi_s(kk)
+                   j=grd%ltosj_s(kk)
+                   work(kk)=grid2(i,j,1)
+                enddo
+             else
+                grid=rwork2d
+                call general_fill_ns(grd,grid,work)
+             endif
+
+          endif
+
+             if ( icount == icm .or. k == nlevs ) then
+                call general_reload(grd,g_z,g_ps,g_tv,g_vor,g_div,g_u,g_v,g_q,g_oz,g_cwmr, &
+                     icount,iflag,ilev,work,uvflag,vordivflag)
+             endif
+
+       enddo ! do k=1,nlevs
+   else ! read_2m
+
+       icount=icount+1
+       iflag(icount)=2
+
+       ! 2m temperature from sfc file
+       if (mype==mype_use(icount)) then
+          call read_vardata(filges, 'tmp2m', rwork2d)
+          
+          if ( diff_res ) then
+             vector(1)=.false.
+             grid_b=rwork2d
+             call fill2_ns(grid_b,grid_c(:,:,1),latb+2,lonb)
+             call g_egrid2agrid(p_high,grid_c,grid2,1,1,vector)
+             do kk=1,grd%itotsub
+                i=grd%ltosi_s(kk)
+                j=grd%ltosj_s(kk)
+                work(kk)=grid2(i,j,1)
+             enddo
+          else
+             grid=rwork2d
+             call general_fill_ns(grd,grid,work)
+          endif
+       endif
+
+       icount=icount + 1
+       iflag(icount)=3
+
+       ! 2m humidity from sfc file
+       if (mype==mype_use(icount)) then
+          call read_vardata(filges, 'spfh2m', rwork2d)
+
+          if ( diff_res ) then
+             vector(1)=.false.
+             grid_b=rwork2d
+             call fill2_ns(grid_b,grid_c(:,:,1),latb+2,lonb)
+             call g_egrid2agrid(p_high,grid_c,grid2,1,1,vector)
+             do kk=1,grd%itotsub
+                i=grd%ltosi_s(kk)
+                j=grd%ltosj_s(kk)
+                work(kk)=grid2(i,j,1)
+             enddo
+          else
+             grid=rwork2d
+             call general_fill_ns(grd,grid,work)
+          endif
+       endif
+
+       icount=icount + 1
+       iflag(icount)=4
```

Comment:
A possible issue here is that I don't understand the assignment of the nodes (i.e., icm). I initially tried to call the new general_2m_reload only if icm==icount, but it wasn't ever called. So I removed the check. It works, but I'm always a bit nervous when I don't fully understand the code. Does this look OK to you? Is it possible that I'm calling general_2m_reload more often than is necessary?

Reply:
Sorry that I am not familiar with this part either. But the increment of `icount `and the assignment of `iflag `does not depend on anything looks like a yellow light to me.

Reply:
> It looks like the assignment of "icount and "iflag" do not depend on whether the "if" block before this executes or not. Is this intended?

If the "if" block is the read_z block, then yes this is intended.

This routine is designed to read either the sfc file (variable: t2m, q2m, sfc press, optionally z - the latter is not done for now) or the atmos file. For the surface file read, if z is requested it will be read that first, then skip the atmos section, then read the variables from the surface file. Either way icount counts the number of variables that have been read, and iflag records the unique number assigned to the icount-th variable.  

I have re-used iflag = 2,3,4 (which have also been assigned to variables in the atmospheric file) but that's OK as the routine reads only one of the sfc or atm files at a time.